torchvision/prototype/datasets/_folder.py

```diff
@@ -0,0 +1,44 @@
+import io
+import os
+import os.path
+import pathlib
+from typing import Union, Tuple, List, Dict, Any
+
+from torch.utils.data import IterDataPipe
+from torch.utils.data.datapipes.iter import FileLister, Concater, FileLoader, Mapper
+
+__all__ = ["from_image_folder"]
+
+
+def _collate(
+    data: Tuple[str, io.IOBase],
+    *,
+    label: int,
+    category: str,
+) -> Dict[str, Any]:
+    path, image_file = data
+    return dict(
+        path=path,
+        image=image_file,
+        label=label,
+        category=category,
+    )
+
+
+def from_image_folder(
+    root: Union[str, pathlib.Path],
+) -> Tuple[IterDataPipe, List[str]]:
+    root = pathlib.Path(root).expanduser().resolve()
+    categories = sorted({item.name for item in os.scandir(root) if item.is_dir})
+    category_dps = []
+    category_dp: IterDataPipe
+    for label, category in enumerate(categories):
+        category_dp = FileLister(str(root / category), recursive=True)
+        category_dp = FileLoader(category_dp)
+        category_dp = Mapper(
+            category_dp,
+            _collate,
+            fn_kwargs=dict(label=label, category=category),
+        )
+        category_dps.append(category_dp)
+    return Concater(*category_dps), categories
```

Comment:
Doesn't this make shuffling much harder, because the elements will be sampled from the same category in order? Or does the shuffler have additional knowledge that knows how to handle those use-cases in a special way?

Reply:
> Doesn't this make shuffling much harder, because the elements will be sampled from the same category in order?

Yes, but I have no idea how to improve this. Maybe we can read one sample from each directory before we start at the top again. Or we implement a `RandomChoser` which picks one element at random from multiple datapipes.

Reply:
Can't we just have a single `FileLister` which is recursive? This way it makes the job of a `Shuffler` much easier, as FileLister could implement a specific property that allows for perfect shuffling for those family of pipes.

@VitalyFedyunin @ejguan thoughts?

Reply:
> Can't we just have a single `FileLister` which is recursive?

We can, yes. As far as I see, that would make sorting harder, since we now need to extract the label and category from each path instead of just setting them. I'll implement it that way and benchmark.

Reply:
If we use

```python
class RandomPicker(IterDataPipe):
    def __init__(self, *datapipes):
        self.datapipes = datapipes

    def __iter__(self):
        non_exhausted = [iter(dp) for dp in self.datapipes]
        while non_exhausted:
            dp = random.choice(non_exhausted)
            try:
                yield next(dp)
            except StopIteration:
                non_exhausted.remove(dp)
```

instead of `Concater` we get a pseudo-shuffled dataset with minimal overhead.